File: picturepicker/src/main/java/com/github2136/picturepicker/other/ImageLoaderInstance.java
package com.github2136.picturepicker.other;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.TextUtils;

/**
 * Created by yb on 2018/5/16.
 */
public class ImageLoaderInstance {
    private ImageLoader mImageLoader;
    private volatile static ImageLoaderInstance ourInstance;

    public static ImageLoaderInstance getInstance(Context context) {
        if (ourInstance == null) {
            synchronized (ImageLoaderInstance.class) {
                if (ourInstance == null) {
                    ourInstance = new ImageLoaderInstance(context);
                }
            }
        }
        return ourInstance;
    }

    private ImageLoaderInstance(Context context) {
        String className = null;
        try {
            ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
            Bundle data = appInfo.metaData;
            if (data != null) {
                for (String key : data.keySet()) {
                    if ("picker_image_loader".equals(appInfo.metaData.get(key))) {
                        className=key;
                    }
                }
//                className = data.getString("");
            }
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        if (!TextUtils.isEmpty(className)) {
            try {
                mImageLoader = (ImageLoader) Class.forName(className).newInstance();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        } else {
            throw new NullPointerException("ImageLoader is null");
        }
    }

    public ImageLoader getImageLoader() {
        return mImageLoader;
    }
}

File: photopicker/src/main/java/com/github2136/photopicker/other/ImageLoaderInstance.kt
package com.github2136.photopicker.other

import android.content.Context
import android.content.pm.PackageManager
import android.text.TextUtils

/**
 * Created by yb on 2018/5/16.
 */
class ImageLoaderInstance private constructor(context: Context) {
    var imageLoader: ImageLoader? = null
        private set

    init {
        var className: String? = null
        try {
            val appInfo = context.packageManager.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
            val data = appInfo.metaData
            if (data != null) {
                for (key in data.keySet()) {
                    if ("picker_image_loader" == appInfo.metaData.get(key)) {
                        className = key
                    }
                }
                //                className = data.getString("");
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        if (!TextUtils.isEmpty(className)) {
            try {
                imageLoader = Class.forName(className!!).newInstance() as ImageLoader
            } catch (e: ClassNotFoundException) {
                e.printStackTrace()
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }

        } else {
            throw NullPointerException("ImageLoader is null")
        }
    }

    companion object {
        @Volatile
        private var ourInstance: ImageLoaderInstance? = null

        fun getInstance(context: Context): ImageLoaderInstance? {
            if (ourInstance == null) {
                synchronized(ImageLoaderInstance::class.java) {
                    if (ourInstance == null) {
                        ourInstance = ImageLoaderInstance(context)
                    }
                }
            }
            return ourInstance
        }
    }
}
